dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataset/DefaultCompleteDataSetRegistrationService.java

```
@@ -113,9 +120,41 @@ public void saveCompleteDataSetRegistration(CompleteDataSetRegistration registra
     notificationEventPublisher.publishEvent(registration);
   }
 
+  /**
+   * Check if a data set is missing data from its compulsory data element operands. DataSet has the
+   * flag compulsoryFieldsCompleteOnly, which when false, actually renders the compulsory elements
+   * not compulsory. That flag is checked first to see if the missing data element operands should
+   * be retrieved. <br>
+   * If there are compulsory elements missing data (and they are compulsory) then an exception is
+   * thrown advising that compulsory elements are required to be filled.
+   *
+   * @param registration registration to check
+   */
+  private void checkCompulsoryDeOperands(CompleteDataSetRegistration registration) {
+    // only get missing compulsory elements if they are actually compulsory
+    if (registration.getDataSet().isCompulsoryFieldsCompleteOnly()) {
+      List<DataElementOperand> missingDataElementOperands =
+          getMissingCompulsoryFields(
+              registration.getDataSet(),
+              registration.getPeriod(),
+              registration.getSource(),
+              registration.getAttributeOptionCombo());
+      if (!missingDataElementOperands.isEmpty()) {
+        String deos =
+            missingDataElementOperands.stream()
+                .map(DataElementOperand::getDisplayName)
+                .collect(Collectors.joining(","));
+        throw new IllegalStateException(
+            "All compulsory data element operands need to be filled: [%s]".formatted(deos));
```

Should we instead throw an exception with an `ErrorCode` here to make it consistent with other validation errors, and allow programmatic client handling? This state is not a "programming error" and could happen e.g. due to user error in an external client.

---

I was about to also leave a comment on error handling but I didn't because I think most if the code for this endpoint/operation does not really follow our more recent goals and I think we better make a PR that moves all the error handling in one PR instead of making it a mixed solution by doing this one in line while other error handling is not.

---

- this error message is consistent with how the mature/bulk API for completing data sets works
  - the bulk API returns a report summary with conflicts (no error code)
  - the single/newish API was implemented to not return any response body/summary
    - it doesn't look like we tried (or thought about) keeping similar behaviour (validation, errors, response etc.) between these APIs for whatever reason
  - if we add an error code in this fix, then behaviour (what little there is) would be inconsistent with the bulk API
- we could change both to return error codes for this particular issue
  - probably best doing that in a separate PR
  - we could improve the single API to return a report also to try to keep things more consistent
  - best not to change behaviour of something else during a bug fix